src/common/cockpitconf.c

```diff
@@ -115,188 +175,202 @@ cockpit_conf_init (void)
-  g_return_val_if_fail (field != NULL, NULL);
+bool
+cockpit_conf_bool (const char *section,
+                   const char *field,
+                   bool defawlt)
+{
+  const char *value = cockpit_conf_string (section, field);
+  if (value)
+    return strcasecmp (value, "yes") == 0 || strcasecmp (value, "true") == 0 || strcmp (value, "1") == 0;
+  return defawlt;
+}
 
-  ensure_cockpit_conf ();
-  key = g_strjoin (delm, section, field, NULL);
+static const char **
+get_strv (Entry *entry, char delimiter)
+{
+  char *stripped, *token;
+  char delimiters[2] = "x";
+  unsigned i;
+
+  if (!entry->value)
+    return NULL;
 
-  if (!g_hash_table_contains (cached_strvs, key))
+  if (entry->strv_cache)
     {
-      value = build_strv (section, field, delm);
-      if (value)
-        g_hash_table_insert (cached_strvs, g_strdup (key), value);
+      if (delimiter != entry->strv_delimiter)
+        errx (1, "cockpitconf: Looking up strv with different delimiters is not supported");
+      return (const char**) entry->strv_cache;
     }
-  else
+
+  /* strip off leading whitespace */
+  for (stripped = entry->value; *stripped && isspace (*stripped); ++stripped)
+  /* strip off trailing whitespace */
+  stripped = strdup (stripped);
+  for (char *c = stripped + strlen (stripped) - 1; c >= stripped && isspace (*c); --c)
+    *c = '\0';
+
+  delimiters[0] = delimiter;
+
+  for (i = 0, token = strtok (stripped, delimiters);
+       token;
+       i++, token = strtok (NULL, delimiters))
     {
-      value = g_hash_table_lookup (cached_strvs, key);
+      entry->strv_cache = reallocarr (entry->strv_cache, i + 2, sizeof (char*));
+      entry->strv_cache[i] = strdup (token);
     }
+  entry->strv_cache[i] = NULL;
 
-  g_free (key);
-  return value;
+  entry->strv_delimiter = delimiter;
+  return (const char**) entry->strv_cache;
+}
+
+const char **
+cockpit_conf_strv (const char *section,
+                   const char *field,
+                   char delimiter)
+{
+  Entry *e = cockpit_conf_lookup (section, field);
+  return e ? get_strv (e, delimiter) : NULL;
 }
 
-guint
-cockpit_conf_guint (const gchar *section,
-                    const gchar *field,
-                    guint default_value,
-                    guint64 max,
-                    guint64 min)
+unsigned
+cockpit_conf_uint (const char *section,
+                   const char *field,
+                   unsigned default_value,
+                   unsigned max,
+                   unsigned min)
 {
-  guint val = default_value;
-  guint64 conf_val;
-  gchar *endptr = NULL;
+  unsigned val = default_value;
+  unsigned long conf_val;
+  char *endptr = NULL;
 
-  const gchar* conf = cockpit_conf_string (section, field);
+  const char* conf = cockpit_conf_string (section, field);
   if (conf)
     {
-      conf_val = g_ascii_strtoull (conf, &endptr, 10);
-      if ((conf_val == G_MAXUINT64 || conf_val == 0) &&
+      conf_val = strtoul (conf, &endptr, 10);
```

Comment:
not your bug, but `errno` needs to be set to `0` before the call.

Reply:
`strtoul()` has really strange handling of negative numbers.  It might make sense to check for a leading `-` before calling it.  I'd recommend just using `strtoll()` here (which will always get you a wider range than the 32bit values in `max` and `min`).

Reply:
Good spot, thanks! (The old code already mis-handled negative numbers). This wouldn't be a major problem in practice as small negative numbers just turn out to be large positive ones and get caught by the "maximum" check, but let's be correct here.